RZAssemblage/RZFilterAssemblage.m

```diff
@@ -0,0 +1,221 @@
+        [self.filteredIndexPaths shiftIndexesStartingAtIndexPath:indexPath by:1];
+
+        id object = [assemblage objectAtIndexPath:indexPath];
+        if ( [self isObjectFiltered:object] == NO ) {
+            [self.changeSet insertAtIndexPath:parentIndexPath];
+        }
+        else {
+            [self.filteredIndexPaths addIndexPath:indexPath];
+        }
+    }
+
+    for ( NSIndexPath *indexPath in changeSet.updatedIndexPaths ) {
+        id object = [assemblage objectAtIndexPath:indexPath];
+
+        NSIndexPath *parentIndexPath = [self exposedIndexPathFromIndexPath:indexPath];
+        if ( [self.filteredIndexPaths containsIndexPath:indexPath] && [self isObjectFiltered:object] == NO ) {
+            [self.filteredIndexPaths removeIndexPath:indexPath];
+
+            [self.changeSet insertAtIndexPath:parentIndexPath];
+        }
+        else if ( [self.filteredIndexPaths containsIndexPath:indexPath] == NO && [self isObjectFiltered:object] ) {
+            [self.filteredIndexPaths addIndexPath:indexPath];
+
+            [self.changeSet removeObject:object atIndexPath:parentIndexPath];
+        }
+        else {
+            [self.changeSet updateAtIndexPath:parentIndexPath];
+        }
+    }
+    [self closeBatchUpdate];
+}
+
+#pragma mark - Private
+
+- (NSIndexPath *)exposedIndexPathFromIndexPath:(NSIndexPath *)indexPath
+{
+    NSIndexPath *parentIndexPath = [NSIndexPath indexPathWithIndexes:NULL length:0];
+    for ( NSUInteger indexPosition = 0; indexPosition < indexPath.length; indexPosition++ ) {
+        NSIndexSet *filtered = [self.filteredIndexPaths indexesAtIndexPath:parentIndexPath];
+        NSUInteger index = [indexPath indexAtPosition:indexPosition];
+        index -= [filtered countOfIndexesInRange:NSMakeRange(0, index)];
+        indexPath = [indexPath rz_indexPathByReplacingIndexAtPosition:indexPosition withIndex:index];
+    }
+    return indexPath;
+}
+
+- (NSIndexPath *)indexPathFromExposedIndexPath:(NSIndexPath *)indexPath
+{
+    NSIndexPath *parentIndexPath = [NSIndexPath indexPathWithIndexes:NULL length:0];
+    for ( NSUInteger indexPosition = 0; indexPosition < indexPath.length; indexPosition++ ) {
+        NSIndexSet *filtered = [self.filteredIndexPaths indexesAtIndexPath:parentIndexPath];
+        __block NSUInteger index = [indexPath indexAtPosition:0];
+        [filtered enumerateRangesUsingBlock:^(NSRange range, BOOL *stop) {
+            if ( index >=  range.location ) {
+                index += range.length;
+            }
+        }];
+        indexPath = [indexPath rz_indexPathByReplacingIndexAtPosition:indexPosition withIndex:index];
+    }
+    return indexPath;
+}
+
+- (BOOL)isObjectFiltered:(id)object
+{
+    BOOL filtered = self.filter && [self.filter evaluateWithObject:object] == NO;
+    return filtered;
+}
+
+@end
+
+@implementation RZAssemblage (Filter)
+
+- (RZFilterAssemblage *)filterAssemblage
+{
+    return [[RZFilterAssemblage alloc] initWithAssemblage:self];
```

Comment:
Honestly I'd rather an `NS_DESIGNATED_INITIALIZER` `initWithAssemblage:filter:`. `filterAssemblage` sounds more like an action (`filteredAssemblage` would sound better), but also doesn't seem like it'd be called often enough to warrant being a convenience category method on `RZAssemblage`